src/components/Modal/src/Modal.vue

```diff
@@ -94,7 +94,10 @@ export default {
 		beforeClose: {
 			immediate: true,
 			handler(hook) {
-				this.modalApi.state.options.beforeCloseHook = hook;
+				if (hook) {
+					// parentModal refers to the layer that this modal is actually in
+					this.modalApi.state.parentModal.state.options.beforeCloseHook = hook;
```

Comment:
Thinking about this some more -- `parentModal` isn't something that is always intended to exist. We dissuade modal stacking for a reason and this will change the original behavior of the beforeClose hook.

I think this needs to be rethought a bit, perhaps having your own custom watcher and setting the data some data with the original hook?

Reply:
Yeah I'm not really a fan of this either, but I tried to do it so the existing functionality would not break, but here's essentially what was happening (which imo was not correct but worked anyway)

- Component opens modal
- Layer `0` creates the modal component and the next modal layer (`1`)
- Modal component attaches the `beforeCloseHook` option to layer `1`
- `modalApi.close` is called, it checks for hooks in layer `1` and continues if valid/no hook
- Layer `1` is torn down

But now, this is how it works:
- Component opens modal with before/after hooks
- Layer `0` creates the modal component, attaches the hooks, and then creates the next layer (`1`)
- The modal component also defines its own hook, and reaches in to set it on layer `0` (this is the `parentModal` stuff that this thread is on).
- `modalApi.close` is called, it now checks for hooks in layer `0`
- Layer `1` is torn down

I think the fundamental complication here is that the `modalApi` within the context of an open modal is actually attached to one layer up (hence the internal changes to reference the hooks from `this.state.options.[before/after]CloseHook` to `vm.currentLayer.state.options.[before/after]CloseHook`).

Reply:
@tatems Really appreciate your write up and considerations around this. I was starting to suspect that there may have been some unintended behavior in the original implementation that "worked" but perhaps in a confusing way. However, I've always interpreted the current behavior as correct due to how we construct the injection of the ModalLayer and Modal having one API injected between both meant the modal is intentionally a child element and provides some level of control over stacking contexts between other modals or other components.

Perhaps there is some rethinking we need to do around all this, but it's probably better for us to internally structure that behavior consistently whether there is modal stacking or not. We made some quick changes to support stacking and somewhat reluctantly. But it does appear that this creates some challenges with the before / after behavior you're trying to achieve.

I'm not exactly sure what that change, if any, that should be. If it's not an immediate need could we dig deeper into some real-world scenarios and we can try to better account for them in both contexts (single modal + stacking modals) -- perhaps planning ahead as another example that we need to consider automated testing to ensure this behavior works consistently with the new hooks.

Reply:
i think i almost got it... and then i lost it... but i know the modal stacking code is very confusing to begin with, so even code that works correctly and as-intended sometimes looks like a confusing hacky mess.

@tatems if you could please add your explanation as a code comment i would appreciate it and will be willing to approve this pr. it's okay if the code comment and is huge and takes up a bunch of lines, the code is confusing to begin with i don't want people to have to dig through old pr comments to have to understand it, would be much nicer to have the explanation and reasoning inline.

Reply:
I think the most confusing part is that modals work just slightly differently enough from popovers and dialogs that it's hard to keep the behaviours straight.

I agree that we should maybe rethink this. And since it's not blocking our immediate project (we only need the dialog functionality), I can drop the modal changes from this and re-open as a separate PR/issue/discussion?